src/ipips/ipip-0412.md

```diff
@@ -0,0 +1,240 @@
+  - https://github.com/ipfs/specs/pull/412
+order: 412
+tags: ['ipips']
+---
+
+## Summary
+
+Adds support for additional, optional content type options that allow the
+client and server to signal or negotiate a specific block order in the returned
+CAR.
+
+## Motivation
+
+We want to make it easier to build light-clients for IPFS. We want them to have
+low memory footprints on arbitrary sized files. The main pain point preventing
+this is the fact that CAR ordering isn't specified.
+
+This require to keeping some kind of reference either on disk, or in memory to
+previously seen blocks for two reasons.
+
+1. Blocks can arrive out of order, meaning when a block is consumed (data is
+   red and returned to the consumer) and when it's received might not match.
+1. Blocks can be reused multiple times, this is handy for cases when you plan
+   to cache on disk but not at all when you want to process a stream with use &
+   forget policy.
+
+What we really want is for the gateway to help us a bit, and give us blocks in
+a useful order.
+
+The existing Trustless Gateway specification does not provide a mechanism for
+negotiating the order of blocks in CAR responses.
+
+This IPIP aims to improve the status quo.
+
+## Detailed design
+
+CAR content type
+([`application/vnd.ipld.car`](https://www.iana.org/assignments/media-types/application/vnd.ipld.car))
+already supports `version` parameter, which allows gateway to indicate which
+CAR flavour is returned with the response.
+
+The proposed solution introduces two new parameters for the content type headers
+in HTTP requests and responses: `order` and `dups`.
+
+The `order` parameter allows the client to indicate its preference for a
+specific block order in the CAR response, and the `dups` parameter specifies
+whether duplicate blocks are allowed in the response.
+
+### Signaling in Request
+
+Content type negotiation is based on section 12.5.1 of :cite[rfc9110].
+
+Clients MAY indicate their preferred block order by sending an `Accept` header in
+the HTTP request. The `Accept` header format is as follows:
+
+```
+Accept: application/vnd.ipld.car; version=1; order=dfs; dups=y
+```
+
+In the future, when more orders or parameters exist, clients will be able to
+specify a list of preferences, for example:
+
+```
+Accept: application/vnd.ipld.car;order=foo, application/vnd.ipld.car;order=dfs;dups=y;q=0.5
+```
+
+The above example is a list of preferences, the client would really like to use
+the hypothetical `order=foo` however if this isn't available it would accept
+`order=dfs` with `dups=y` instead (lower priority indicated via `q` parameter,
+as noted in :cite[rfc9110]).
+
+#### `order` CAR content type parameter
+
+The `order` parameter accepts the following values:
+
+- `dfs`: [Depth-First Search](https://en.wikipedia.org/wiki/Depth-first_search)
+  order, allows for streaming responses with minimal memory usage
+- `rnd`: Unknown (random) order, the implicit default when `order` parameter is missing.
```

Comment:
```suggestion
- `dfs`: [Depth-First Search](https://en.wikipedia.org/wiki/Depth-first_search)
  order, allows for streaming responses with minimal memory usage
- `weak`: weak maintains a seen set, the seen set is initialized with the CID passed in, the gateway is only allowed to send blocks that are in the seen set. When a block is sent it's childrens that are relevent to the content of the request are added to the seen set, if `dups` is `n` the block that was sent is removed from the seen set.
  In other words this provide vertical ordering but not horizontal ordering.
  This ensures the gateway never send a block the client can't know about yet and ensure the stream is incrementally verifiable while providing a weaker ordering requirement which allows the gateway implementation more creative potentially faster strategies.
- `rnd`: Unknown (random) order, the implicit default when `order` parameter is missing.
```

Reply:
`rnd` doesn't mean random, correct, but just arbitrary/unspecified?
I assume we're not requiring a gateway to shuffle things by default.

Reply:
I think something like `unknown` is better than `rnd` as a name here. I'm also curious about the interest in support for `unknown/rnd` given that it's not incrementally verifiable and therefore necessitates both added trust and added (potentially unbounded) buffering.

@Jorropo IIUC what you are describing with `weak` is more succinctly "any valid graph traversal order", since the types of orders you're trying to exclude are ones like putting all the leaves first and the root at the end. Is that right?

Reply:
-1 to adding additional items to this list in the first pass, the more overengineered this spec is in the first instance the less likely it is to be properly, or fully implemented in >1, or even >0 places. This stuff is so far from nontrivial and the cross-language barriers are huge and we haven't even crossed that barrier for simple depth-first yet.

Reply:
and also let's keep it simple like @rvagg says in first pass. We can add more later.

Reply:
@rvagg @hannahhoward we already have implementations that makes download faster if using `weak` instead of `dfs` in `ipfs pin add` and RAPIDE.
I don't have the unixfs decoder since we havn't deployed thoses on gateway yet, this is about an afternoon of work (have a map that keeps tracks of all open ranges in the file, when a new block is downloaded progress open spaces and use `.WriteAt` to write to the correct offset).

`dfs` is already a valid implementation of `weak` so this is not represent any real cost for implementations since we already want to support `dfs`. Breath first search is also a valid implementation, anything that sends parents before it sends children is a valid implementation.

Reply:
Nothing prevents us from adding `weak` later. But any "arbitrarily valid traversal" order feels ill defined for me to be going in a formal IPFS spec, at this point, in order to support prototype clients.

I'd like to request additional discussion and a seperate spec, similar to how this spec evolved from 402.

Reply:
@hannahhoward are you only objecting to `weak` or also `unknown/rnd`?

Reply:
just weak. I don't care about parameter names. 

Nothing prevents someone from passing `weak` any more than the range request standard allows you specify other values besides `bytes`. I'm not saying it's a bad idea or a useless ordering. I'm just saying from a standards point of view, we should be putting in values that have deployment in prod and an existing product using it. such that the standard is just bringing things in line with how production is already working, and potentially enabling others to use it. we shouldn't generally design standards around possible use cases. my argument is if @Jorropo wants to keep working on `weak` and develop some deployments, and propose it's a recognized value for others to use, go for it. But right now we're aiming for the minimal standard to get Rhea over the line.

Reply:
- I've renamed `rnd` to `unk` (unknown) and clarified purpose in 50223b296708b90fc610fb3f55387b0d8117d9d2
  - We don't expect to see `unk`—it is there to document the implicit behavior when `order` is not explicitly stated
- Agree we want to keep the scope of this IPIP surgical and limited to what we already have and use in the ecosystem (`unk` and `dfs`)
- @Jorropo feel free to propose and discuss more orderings in separate IPIPs  -- `weak` is a good candidate, but needs working code first AND fixtures that are different than `dfs`, and we don't want to slow down this IPIP by the details of fleshing it out.

Reply:
If I'm understanding correctly, we have something similar to @Jorropo's "weak" ordering in our [dagula](https://github.com/web3-storage/dagula) library for exporting DAGs:

It just fetches blocks and sends them as soon as it knows about them. The interesting part is that you'll not receive a block you don't already know that you will need, so it's verifiable in that sense, but they might not arrive quite in the order you need them.

It makes the server side very fast and memory efficient because you don't keep any state or buffer and don't need to recurse. The trade off is that the _receipient_ has to re-order and buffer any out of order blocks.

I believe implementing depth first ordering would make that export slower and/or more memory intensive...so it might be nice to have something inbetween `dfs` and `rdn` that is verifiable _and_ fast.

In our case we have memory constrained workers running in the cloud that are serving blocks to clients that probably have significantly more memory/ability to buffer than the worker does. They also probably want to receive data very quickly (who doesn't right?).

Another factor to consider is that our workers fetch blocks from CAR files stored in a bucket, with RTTs that are significantly greater than an SSD. Fetching blocks 1 by 1 to order as DFS (without blowing out memory) is going to be really slow.

So I guess this comment is to +1 on the "weak" ordering, and to say there _is_ a prod implementation of this (we use Dagula in our gateway [Freeway](https://github.com/web3-storage/freeway)).

Reply:
Let's keep the scope of this IPIP limited to opt-in `order=dfs` and `dups=y|n`.

After this IPIP lands, we can fill a new one to add specification for `weak` or any other ordering.

For now, if you can't support `order=dfs`, or `dups`, just don't set these parameters on CAR responses.